File: examples/loan_approval.js
var $state = {
  
  yearsInJob: 0,
  
  approvalMessage: "",
  
  interestRate: 0,
  
  employed: false,
  
  totalInterest: 0,
  
  hasCoSigner: false,
  
  not_approvalMessage: "Not Approved",
  
  jobTitle: "",
  
  loanTerm: 0,
  
  age: 0,
  
  hasLicense: false,
  
  loanApproved: false,
  
  loanAmount: 0,
  
  healthcareExpenses: 0,
  
  monthlyExpenses: 0,
  
  income: 0,
  
  retired: false,
  
  annualPension: 0,
  
  totalRepayment: 0,
  
  netPension: 0,
  
  seniorDiscount: false,
  
  monthlyDebts: 0,
  
  yearsRetired: 0,
  
  monthlySalary: 0,
  
  lookingForJob: false,
  
  grade: 0,
  
  inSchool: false,
  
  monthlyPayment: 0,
  
  fullName: "",
  
  annualSavings: 0,
  
}
function $update(name, value) {
   let change = false;
   let newVal = undefined;
   let div = undefined;
   if (name !== undefined) {
      $state[name] = value;
   }
   do {
     change = false;
     
     div = document.getElementById('fullName_div_25');
     div.style.display = true ? 'block' : 'none'; 
     
     
     div = document.getElementById('age_div_70');
     div.style.display = true ? 'block' : 'none'; 
     
     
     div = document.getElementById('hasLicense_div_104');
     div.style.display = true ? 'block' : 'none'; 
     
     
     div = document.getElementById('inSchool_div_183');
     div.style.display = (true && ($state.age < 18)) ? 'block' : 'none'; 
     
     
     div = document.getElementById('grade_div_258');
     div.style.display = ((true && ($state.age < 18)) && $state.inSchool) ? 'block' : 'none'; 
     
     
     div = document.getElementById('employed_div_360');
     div.style.display = ((true && !((($state.age < 18)))) && (($state.age >= 18) && ($state.age <= 65))) ? 'block' : 'none'; 
     
     
     div = document.getElementById('jobTitle_div_445');
     div.style.display = (((true && !((($state.age < 18)))) && (($state.age >= 18) && ($state.age <= 65))) && $state.employed) ? 'block' : 'none'; 
     
     
     div = document.getElementById('yearsInJob_div_496');
     div.style.display = (((true && !((($state.age < 18)))) && (($state.age >= 18) && ($state.age <= 65))) && $state.employed) ? 'block' : 'none'; 
     
     
     div = document.getElementById('monthlySalary_div_578');
     div.style.display = (((true && !((($state.age < 18)))) && (($state.age >= 18) && ($state.age <= 65))) && $state.employed) ? 'block' : 'none'; 
     
     
     div = document.getElementById('monthlyExpenses_div_640');
     div.style.display = (((true && !((($state.age < 18)))) && (($state.age >= 18) && ($state.age <= 65))) && $state.employed) ? 'block' : 'none'; 
     
     
     div = document.getElementById('annualSavings_div_713');
     div.style.display = (((true && !((($state.age < 18)))) && (($state.age >= 18) && ($state.age <= 65))) && $state.employed) ? 'block' : 'none'; 
     
     newVal = ((($state.monthlySalary - $state.monthlyExpenses)) * 12);
     if (newVal !== $state.annualSavings) {
       let elt = document.getElementById('annualSavings_widget_713');
       $state.annualSavings = newVal;
       
       elt.value = newVal;
       
       change = true;
     }
     
     
     div = document.getElementById('lookingForJob_div_822');
     div.style.display = (((true && !((($state.age < 18)))) && (($state.age >= 18) && ($state.age <= 65))) && !(($state.employed))) ? 'block' : 'none'; 
     
     
     div = document.getElementById('retired_div_910');
     div.style.display = ((true && !((($state.age < 18)))) && !(((($state.age >= 18) && ($state.age <= 65))))) ? 'block' : 'none'; 
     
     
     div = document.getElementById('yearsRetired_div_975');
     div.style.display = (((true && !((($state.age < 18)))) && !(((($state.age >= 18) && ($state.age <= 65))))) && $state.retired) ? 'block' : 'none'; 
     
     
     div = document.getElementById('annualPension_div_1039');
     div.style.display = (((true && !((($state.age < 18)))) && !(((($state.age >= 18) && ($state.age <= 65))))) && $state.retired) ? 'block' : 'none'; 
     
     
     div = document.getElementById('healthcareExpenses_div_1101');
     div.style.display = (((true && !((($state.age < 18)))) && !(((($state.age >= 18) && ($state.age <= 65))))) && $state.retired) ? 'block' : 'none'; 
     
     
     div = document.getElementById('netPension_div_1180');
     div.style.display = (((true && !((($state.age < 18)))) && !(((($state.age >= 18) && ($state.age <= 65))))) && $state.retired) ? 'block' : 'none'; 
     
     newVal = ($state.annualPension - $state.healthcareExpenses);
     if (newVal !== $state.netPension) {
       let elt = document.getElementById('netPension_widget_1180');
       $state.netPension = newVal;
       
       elt.value = newVal;
       
       change = true;
     }
     
     
     div = document.getElementById('seniorDiscount_div_1296');
     div.style.display = true ? 'block' : 'none'; 
     
     newVal = (($state.age >= 65) && $state.retired);
     if (newVal !== $state.seniorDiscount) {
       let elt = document.getElementById('seniorDiscount_widget_1296');
       $state.seniorDiscount = newVal;
       
       elt.checked = newVal;
       
       change = true;
     }
     
     
     div = document.getElementById('income_div_1398');
     div.style.display = true ? 'block' : 'none'; 
     
     
     div = document.getElementById('monthlyDebts_div_1446');
     div.style.display = true ? 'block' : 'none'; 
     
     
     div = document.getElementById('hasCoSigner_div_1501');
     div.style.display = true ? 'block' : 'none'; 
     
     
     div = document.getElementById('loanAmount_div_1564');
     div.style.display = true ? 'block' : 'none'; 
     
     
     div = document.getElementById('interestRate_div_1608');
     div.style.display = true ? 'block' : 'none'; 
     
     
     div = document.getElementById('loanTerm_div_1666');
     div.style.display = true ? 'block' : 'none'; 
     
     
     div = document.getElementById('totalInterest_div_1709');
     div.style.display = true ? 'block' : 'none'; 
     
     newVal = (function () { const div = 100; return div !== 0 ? (((($state.loanAmount * $state.interestRate) * $state.loanTerm)) / div) : 0; })();
     if (newVal !== $state.totalInterest) {
       let elt = document.getElementById('totalInterest_widget_1709');
       $state.totalInterest = newVal;
       
       elt.value = newVal;
       
       change = true;
     }
     
     
     div = document.getElementById('totalRepayment_div_1809');
     div.style.display = true ? 'block' : 'none'; 
     
     newVal = ($state.loanAmount + $state.totalInterest);
     if (newVal !== $state.totalRepayment) {
       let elt = document.getElementById('totalRepayment_widget_1809');
       $state.totalRepayment = newVal;
       
       elt.value = newVal;
       
       change = true;
     }
     
     
     div = document.getElementById('monthlyPayment_div_1892');
     div.style.display = true ? 'block' : 'none'; 
     
     newVal = (function () { const div = (($state.loanTerm * 12)); return div !== 0 ? ($state.totalRepayment / div) : 0; })();
     if (newVal !== $state.monthlyPayment) {
       let elt = document.getElementById('monthlyPayment_widget_1892');
       $state.monthlyPayment = newVal;
       
       elt.value = newVal;
       
       change = true;
     }
     
     
     div = document.getElementById('loanApproved_div_1984');
     div.style.display = true ? 'block' : 'none'; 
     
     newVal = (((($state.income > 50000) && ($state.monthlyDebts < 10000))) || $state.hasCoSigner);
     if (newVal !== $state.loanApproved) {
       let elt = document.getElementById('loanApproved_widget_1984');
       $state.loanApproved = newVal;
       
       elt.checked = newVal;
       
       change = true;
     }
     
     
     div = document.getElementById('approvalMessage_div_2114');
     div.style.display = (true && $state.loanApproved) ? 'block' : 'none'; 
     
     newVal = "Approved";
     if (newVal !== $state.approvalMessage) {
       let elt = document.getElementById('approvalMessage_widget_2114');
       $state.approvalMessage = newVal;
       
       elt.value = newVal;
       
       change = true;
     }
     
     
     div = document.getElementById('not_approvalMessage_div_2214');
     div.style.display = (true && !(($state.loanApproved))) ? 'block' : 'none'; 
     
     newVal = "Not Approved";
     if (newVal !== $state.not_approvalMessage) {
       let elt = document.getElementById('not_approvalMessage_widget_2214');
       $state.not_approvalMessage = newVal;
       
       elt.value = newVal;
       
       change = true;
     }
     
     
   } while (change);
}

File: examples/loan_approval.py
def loan_approval_myql_oracle(driver):
    state = driver.execute_script('return $state;');
    
    assert not True or driver.find_elements(By.ID, 'fullName_div_25')[0].is_displayed
    assert not True or driver.find_elements(By.ID, 'age_div_70')[0].is_displayed
    assert not True or driver.find_elements(By.ID, 'hasLicense_div_104')[0].is_displayed
    assert not (True and (state['age'] < 18)) or driver.find_elements(By.ID, 'inSchool_div_183')[0].is_displayed
    assert not ((True and (state['age'] < 18)) and state['inSchool']) or driver.find_elements(By.ID, 'grade_div_258')[0].is_displayed
    assert not ((True and not (((state['age'] < 18)))) and ((state['age'] >= 18) and (state['age'] <= 65))) or driver.find_elements(By.ID, 'employed_div_360')[0].is_displayed
    assert not (((True and not (((state['age'] < 18)))) and ((state['age'] >= 18) and (state['age'] <= 65))) and state['employed']) or driver.find_elements(By.ID, 'jobTitle_div_445')[0].is_displayed
    assert not (((True and not (((state['age'] < 18)))) and ((state['age'] >= 18) and (state['age'] <= 65))) and state['employed']) or driver.find_elements(By.ID, 'yearsInJob_div_496')[0].is_displayed
    assert not (((True and not (((state['age'] < 18)))) and ((state['age'] >= 18) and (state['age'] <= 65))) and state['employed']) or driver.find_elements(By.ID, 'monthlySalary_div_578')[0].is_displayed
    assert not (((True and not (((state['age'] < 18)))) and ((state['age'] >= 18) and (state['age'] <= 65))) and state['employed']) or driver.find_elements(By.ID, 'monthlyExpenses_div_640')[0].is_displayed
    assert not (((True and not (((state['age'] < 18)))) and ((state['age'] >= 18) and (state['age'] <= 65))) and state['employed']) or driver.find_elements(By.ID, 'annualSavings_div_713')[0].is_displayed
    assert state['annualSavings'] == (((state['monthlySalary'] - state['monthlyExpenses'])) * 12)
    assert not (((True and not (((state['age'] < 18)))) and ((state['age'] >= 18) and (state['age'] <= 65))) and not ((state['employed']))) or driver.find_elements(By.ID, 'lookingForJob_div_822')[0].is_displayed
    assert not ((True and not (((state['age'] < 18)))) and not ((((state['age'] >= 18) and (state['age'] <= 65))))) or driver.find_elements(By.ID, 'retired_div_910')[0].is_displayed
    assert not (((True and not (((state['age'] < 18)))) and not ((((state['age'] >= 18) and (state['age'] <= 65))))) and state['retired']) or driver.find_elements(By.ID, 'yearsRetired_div_975')[0].is_displayed
    assert not (((True and not (((state['age'] < 18)))) and not ((((state['age'] >= 18) and (state['age'] <= 65))))) and state['retired']) or driver.find_elements(By.ID, 'annualPension_div_1039')[0].is_displayed
    assert not (((True and not (((state['age'] < 18)))) and not ((((state['age'] >= 18) and (state['age'] <= 65))))) and state['retired']) or driver.find_elements(By.ID, 'healthcareExpenses_div_1101')[0].is_displayed
    assert not (((True and not (((state['age'] < 18)))) and not ((((state['age'] >= 18) and (state['age'] <= 65))))) and state['retired']) or driver.find_elements(By.ID, 'netPension_div_1180')[0].is_displayed
    assert state['netPension'] == (state['annualPension'] - state['healthcareExpenses'])
    assert not True or driver.find_elements(By.ID, 'seniorDiscount_div_1296')[0].is_displayed
    assert state['seniorDiscount'] == ((state['age'] >= 65) and state['retired'])
    assert not True or driver.find_elements(By.ID, 'income_div_1398')[0].is_displayed
    assert not True or driver.find_elements(By.ID, 'monthlyDebts_div_1446')[0].is_displayed
    assert not True or driver.find_elements(By.ID, 'hasCoSigner_div_1501')[0].is_displayed
    assert not True or driver.find_elements(By.ID, 'loanAmount_div_1564')[0].is_displayed
    assert not True or driver.find_elements(By.ID, 'interestRate_div_1608')[0].is_displayed
    assert not True or driver.find_elements(By.ID, 'loanTerm_div_1666')[0].is_displayed
    assert not True or driver.find_elements(By.ID, 'totalInterest_div_1709')[0].is_displayed
    assert state['totalInterest'] == (lambda div: ((((state['loanAmount'] * state['interestRate']) * state['loanTerm'])) / div) if div != 0 else 0)(100)
    assert not True or driver.find_elements(By.ID, 'totalRepayment_div_1809')[0].is_displayed
    assert state['totalRepayment'] == (state['loanAmount'] + state['totalInterest'])
    assert not True or driver.find_elements(By.ID, 'monthlyPayment_div_1892')[0].is_displayed
    assert state['monthlyPayment'] == (lambda div: (state['totalRepayment'] / div) if div != 0 else 0)(((state['loanTerm'] * 12)))
    assert not True or driver.find_elements(By.ID, 'loanApproved_div_1984')[0].is_displayed
    assert state['loanApproved'] == ((((state['income'] > 50000) and (state['monthlyDebts'] < 10000))) or state['hasCoSigner'])
    assert not (True and state['loanApproved']) or driver.find_elements(By.ID, 'approvalMessage_div_2114')[0].is_displayed
    assert state['approvalMessage'] == 'Approved'
    assert not (True and not ((state['loanApproved']))) or driver.find_elements(By.ID, 'not_approvalMessage_div_2214')[0].is_displayed
    assert state['not_approvalMessage'] == 'Not Approved'
